src/gui_common/AddWindowReorderingSupportToSiblings.cs

```diff
@@ -0,0 +1,328 @@
+        Setup();
+
+        if (recursive)
+        {
+            // Connection is recursive, connect to other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.ConnectWindow(
+                    window, windowReorderingAncestors[i].Sibling, recursive);
+            }
+        }
+
+        window.Connect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        connectedWindows.Add(window, topNode);
+        connectedSiblings.Add(topNode);
+
+        // Update top sibling
+        if (topSibling is null || topSibling.GetIndex() < topNode.GetIndex())
+            topSibling = topNode;
+    }
+
+    public void DisconnectWindow(CustomWindow window, bool recursive)
+    {
+        if (recursive)
+        {
+            // Connection was recursive, disconnect from other window reordering nodes
+            for (int i = 0; i < windowReorderingAncestors.Count; i++)
+            {
+                windowReorderingAncestors[i].Node.DisconnectWindow(window, recursive);
+            }
+        }
+
+        window.Disconnect(nameof(Dragged), this, nameof(OnWindowReorder));
+
+        var windowSibling = connectedWindows[window];
+        connectedWindows.Remove(window);
+
+        // Find if another connected window has the same sibling
+        bool foundSibling = false;
+        foreach (var sibling in connectedSiblings)
+        {
+            if (sibling == windowSibling)
+            {
+                foundSibling = true;
+                break;
+            }
+        }
+
+        if (!foundSibling)
+        {
+            // No other window has the same sibling so it can be removed
+            connectedSiblings.Remove(windowSibling);
+        }
+    }
+
+    /// <summary>
+    ///   Finds whose windows ancestor is currently on the top
+    /// </summary>
+    public void UpdateTopSibling()
+    {
+        if (topSibling is not null)
+        {
+            // Top sibling is already current
+            return;
+        }
+
+        var siblings = parent.GetChildren();
+
+        // Search for the first sibling that is part of the reordering system
+        for (int i = siblings.Count - 1; i >= 0; i--)
+        {
+            var sibling = (Node)siblings[i];
+            if (!connectedSiblings.Contains(sibling))
+                continue;
+
+            topSibling = sibling;
+            break;
+        }
+    }
+
+    /// <summary>
+    ///   Reoders a window by setting its ancestor to the position of current top window, making it appear on top of
+    ///   others.
+    /// </summary>
+    public void OnWindowReorder(CustomDialog window)
+    {
+        // Get a sibling that is an ancestor of this window
+        Node targetSibling = connectedWindows[window];
+
+        UpdateTopSibling();
+
+        int topSiblingIndex = topSibling!.GetIndex();
```

Comment:
It's possible for UpdateTopSibling to not find anything so an explicit null check and exception with an error message should be performed on `topSibling`.

Reply:
I don't see how this can happen. Reason: there has to be at least one sibling (an ancestor of the window that asked to be reordered). So I think that this check is not useful here. Or do you think that I am wrong?

Reply:
Well this is a public method, so someone could accidentally call this with an unrelated CustomDialog object... so a bit of defensive programming could be used here to throw a specific exception with an error message in that problem situation.

Reply:
This and OnWindowReorder shoud be private. I will fix it and well, it won't hurt if I add the exception here.